src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/service/VisitSchedulerSessionsService.kt

```diff
@@ -0,0 +1,164 @@
+import uk.gov.justice.digital.hmpps.hmppsmanageprisonvisitsorchestration.dto.visit.scheduler.SessionCapacityDto
+import uk.gov.justice.digital.hmpps.hmppsmanageprisonvisitsorchestration.dto.visit.scheduler.SessionScheduleDto
+import uk.gov.justice.digital.hmpps.hmppsmanageprisonvisitsorchestration.dto.visit.scheduler.VisitSessionDto
+import uk.gov.justice.digital.hmpps.hmppsmanageprisonvisitsorchestration.dto.visit.scheduler.enums.SessionRestriction
+import uk.gov.justice.digital.hmpps.hmppsmanageprisonvisitsorchestration.dto.visit.scheduler.enums.SessionRestriction.CLOSED
+import uk.gov.justice.digital.hmpps.hmppsmanageprisonvisitsorchestration.dto.visit.scheduler.enums.SessionRestriction.OPEN
+import uk.gov.justice.digital.hmpps.hmppsmanageprisonvisitsorchestration.dto.whereabouts.ScheduledEventDto
+import uk.gov.justice.digital.hmpps.hmppsmanageprisonvisitsorchestration.exception.DateRangeNotFoundException
+import java.time.LocalDate
+import java.time.LocalTime
+
+@Service
+class VisitSchedulerSessionsService(
+  private val visitSchedulerClient: VisitSchedulerClient,
+  private val appointmentsService: AppointmentsService,
+  private val prisonerProfileService: PrisonerProfileService,
+  private val prisonService: PrisonService,
+) {
+  companion object {
+    val LOG: Logger = LoggerFactory.getLogger(this::class.java)
+    const val CLASHING_APPOINTMENT_LOG_MSG =
+      "Visit session for prisonerId - {}, date - {}, start time - {}, end time - {} is unavailable as it clashes with {} medical / legal appointment(s), appointment details - {}"
+  }
+
+  fun getVisitSessions(prisonCode: String, prisonerId: String?, min: Int?, max: Int?): List<VisitSessionDto>? {
+    return visitSchedulerClient.getVisitSessions(prisonCode, prisonerId, min, max)
+  }
+
+  fun getAvailableVisitSessions(
+    prisonCode: String,
+    prisonerId: String,
+    requestedSessionRestriction: SessionRestriction?,
+    visitors: List<Long>?,
+    withAppointmentsCheck: Boolean,
+  ): List<AvailableVisitSessionDto> {
+    val sessionRestriction = updateRequestedRestriction(requestedSessionRestriction, prisonerId, visitors)
+
+    val dateRange = prisonService.getToDaysBookableDateRange(prisonCode)
+    var availableVisitSessions = try {
+      val updatedDateRange =
+        visitors?.let { prisonerProfileService.getBannedRestrictionDateRage(prisonerId, visitors, dateRange) }
+          ?: dateRange
+      visitSchedulerClient.getAvailableVisitSessions(prisonCode, prisonerId, sessionRestriction, updatedDateRange)
+    } catch (e: DateRangeNotFoundException) {
+      LOG.error("getAvailableVisitSessions range is not returned therefore we do not have a valid date range and should return an empty list")
+      emptyList()
+    }
+
+    availableVisitSessions = if (withAppointmentsCheck && availableVisitSessions.isNotEmpty()) {
+      val higherPriorityAppointments = appointmentsService.getHigherPriorityAppointments(prisonerId, dateRange.fromDate, dateRange.toDate)
```

Comment:
any chance you could lift this out and put it in a sub method

Reply:
Updated